### PR TITLE
Added --compile-fast to compile with tcc.

### PR DESCRIPTION
**Pull request by jacereda**
Not sure if this is the best approach. Maybe we should just read a `CC` environment variable?

Timings on some very old hardware:

- `CARP_OPTS=--compile-fast ./run_carp_tests.sh --no_sdl`: 2m:51
- `./run_carp_tests.hs --no_sdl`: 8m58

### CODE REVIEW

**Comment by jacereda** [comment]
Forgot to say that this also fixes non-standard code generated for `(deftype Simple [])`. C doesn't allow empty structs, even if some compilers allow it. It just generates a `char __dummy` field. Should also help with Visual C.

File: core/carp_long.h [hunk 1]
```diff
@@ -10,7 +10,23 @@ Long Long__MUL_(Long x, Long y) {
...
+}
+bool Long_safe_MINUS_sub(Long x, Long y, Long* res) {
+    Long r = x - y;
+    *res = r;
+    return (y > 0 && x < (INT64_MIN + y)) || (y < 0 && x > (INT64_MAX + y));
+}
+bool Long_safe_MINUS_mul(Long x, Long y, Long* res) {
+    Long r = x * y;
+    *res = r;
+    return (r / y) != x;
```
Comment (hellerve) [inline core/carp_long.h:27]: Won’t this explode if `y` is `0`?

Comment (jacereda) [reply core/carp_long.h]: Doh! Fixed, thanks!

I also fixed the timings in the description, they were reversed.

**Comment by hellerve** [comment]
This seems a bit like a “miscellaneous" PR. Are all of the features still relevant? I think there was some discussion about changing the TCC bit in #774?

**Comment by jacereda** [comment]
Well, besides adding the tcc stuff it also increases the chances it will compile in non-clang compilers. I'll send another PR to add the `--cc` flag but this one should probably be rebased on top to provide the empty struct fix and the safe ops fallback.

**Comment by jacereda** [comment]
@eriksvedang I think the windows build might be a problem with the CI machine. I just built on Windows with that same resolver.

**Comment by GrayJack** [comment]
I do think it would be a better idea to have a `--cc` or read the `CC` environment variable

Or maybe both? Default to `CC` but if `--cc` specified, use what was specified

**Comment by hellerve** [comment]
Yes, we do. I think it might make sense to also give an official precedence/semantics to potentially over-lapping configurations like this. I’d propose:

```
; right overrides left
Global config file -> CLI flags -> code
```

What do y’all think?

**Comment by GrayJack** [comment]
I think that would be the best option :smile:

**Comment by jacereda** [comment]
IMHO CLI flags should have preference. Say I want to cross-compile or use tcc. I don't want to edit the code or config file. It would be better to just pass `-cc tcc`.

Having said that, maybe something like this would be quite flexible:

```
carp --eval-postload '(Project.config "compiler" "tcc")' -x whatever.carp
```

**Comment by jacereda** [comment]
I mean, we could also have `--eval-preload` and you would have:
```
Config file -> preload code -> code -> postload code
```

**Comment by scolsen** [comment]
I agree that CLI flags should have the highest precedence. pre and post eval is cool, but I think somewhat verbose/complicated for this particular feature—it feels more natural to me to pass a compiler choice as a a simple flag value—the pre-post-eval mechanism might be very useful generally speaking, though.

**Comment by jacereda** [comment]
You're right, it feels a bit verbose, but maybe it's not so common to deserve a shorthand? You can always  create, say, a `fast-carp.sh` that does:

```
./carp.sh --eval-preload '(Project.config "compiler" "tcc")' "$@"
```

The part that worries me is that you might want to override in the command line any of the current or future Project settings and we could end up having tons of switches.

**Comment by scolsen** [comment]
That’s a fair point! Sounds good—are we going to make it just a general mechanism that evals some arbitrary form, or are we going to restrict it to evaluating only `Project.config` forms?

**Pull request merged**